src/devices/r900.c

```diff
@@ -0,0 +1,297 @@
+    if ( bit_1 == 'x' || bit_2 == 'x' )
+        return -99;
+    sum = 6 * ((int)bit_1 - 48) + ((int)bit_2 - 48);
+    return sum;
+}
+
+//
+// parse values from the reconstructed bitstream
+//
+static uint32_t parse_value( char *bb, int start, int end ) {
+    char value[33] = {0}; // no values larger than 4 bytes
+
+    for ( int i=start; i < end; i++ ) {
+        value[i-start] = bb[i];
+    }
+    value[end-start] = 0;
+    return strtol(value, NULL, 2);
+}
+
+static int r900_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{   // partial preamble and sync word shifted by 1 bit
+    uint8_t const preamble[] = {0x55, 0x55, 0x55, 0xa9, 0x66, 0x69, 0x65};
+    uint16_t preamble_bits = sizeof(preamble) * 8;
+
+    uint32_t meter_id;
+    uint32_t Unkn1;
+    uint32_t NoUse;
+    uint32_t BackFlow;
+    uint32_t Consumption;
+    uint32_t Unkn3;
+    uint32_t Leak;
+    uint32_t LeakNow;
+    int row = 0;
+
+    // Search for preamble and sync-word
+    unsigned start_pos = bitbuffer_search(bitbuffer, row, 0, preamble, preamble_bits);
+    // No preamble detected
+    if (start_pos == bitbuffer->bits_per_row[row])
+        return DECODE_ABORT_EARLY;
+    decoder_logf(decoder, 1, __func__, "r900 protocol detected, buffer is %d bits length", bitbuffer->bits_per_row[row]);
+    
+    // Remove preamble and sync word, keep whole payload
+    uint8_t b[21];
+    bitbuffer_extract_bytes(bitbuffer, row, start_pos + preamble_bits, b, 21 * 8);
+
+    // check that (bitbuffer->bits_per_row[0]) greater than (start_pos+sizeof(preamble)*8+168)
+    if (start_pos+preamble_bits+168 > bitbuffer->bits_per_row[0])
+        return DECODE_ABORT_EARLY;
+
+    uint8_t *bb = bitbuffer->bb[0];
+    int base6_dec[21];
+    char final_bits[106] = {0}; // 21 * 5 = 105 bits of data + 0
+    int count = 0;
+
+    /*
+     * Each group of four of these chips must be interpretted as a digit in base 6 
+	 *             according to the following mapping:
+     * 0011 -> 0
+     * 0101 -> 1
+     * 0110 -> 2
+     * 1100 -> 3
+     * 1010 -> 4
+     * 1001 -> 5
+    */
+    // create a pair of char bit array of '0' and '1' for each base6 byte
+    for (uint16_t k = start_pos+preamble_bits; k < start_pos + preamble_bits + 168; k=k+8) {
+        char chip_byte[9] = {0};
+        for ( int j = 0; j < 8; j++) {
+            int bit = bitrow_get_bit(bb, k+j);
+            if (bit == 1)
+                chip_byte[j] = '1';
+            else
+                chip_byte[j] = '0';
+        }
+
+        // convert the above to a base6 integer
+        base6_dec[count] = get_base6_dec( chip_byte );
+        if ( base6_dec[count] == -99 )
+            return DECODE_ABORT_EARLY;
+
+        count++;
+    }
+
+    // convert the base6 integers above into binary bits for decoding data
+    for ( int i=0; i < 21; i++ ) {
+        strncat(final_bits, bit_rep[base6_dec[i]], 5);
+    }
+
+    // decode the data
+    //
+    // meter_id 32 bits
+    meter_id = parse_value( final_bits, 0, 32 );
```

Comment:
and here read from a binary buffer not a string, the decode above should produce a bitrow not a string.